src/core/exprfilter.cpp

```diff
@@ -18,540 +18,2481 @@
+            return true;
+
+        std::sort(terms.begin(), terms.end(), pred);
+        return false;
+    }
+
+    ExpressionTreeNode *emit(ExpressionTree &tree, const ValueIndex &index) const
+    {
+        ExpressionTreeNode *head = nullptr;
+
+        for (const auto &term : terms) {
+            ExpressionTreeNode *node = term.emit(tree, index);
+
+            if (head) {
+                ExpressionTreeNode *addNode = tree.makeNode(term.getCoeff() < 0 ? ExprOpType::SUB : ExprOpType::ADD);
+                addNode->setLeft(head);
+                addNode->setRight(node);
+                head = addNode;
+            } else {
+                head = node;
+            }
+        }
+
+        if (head) {
+            ExpressionTreeNode *addNode = tree.makeNode(scalarTerm < 0 ? ExprOpType::SUB : ExprOpType::ADD);
+            addNode->setLeft(head);
+            addNode->setRight(tree.makeNode({ ExprOpType::CONSTANT, std::fabs(scalarTerm) }));
+            head = addNode;
+        } else {
+            head = tree.makeNode({ ExprOpType::CONSTANT, 0.0f });
+        }
+
+        return head;
+    }
+};
+
+bool analyzeAdditiveExpression(ExpressionTree &tree, ExpressionTreeNode &node)
+{
+    size_t origNumTerms = 0;
+    AdditiveSequence expr;
+    ValueIndex index;
+
+    node.preorder([&](ExpressionTreeNode &node)
+    {
+        if (isOpCode(node, { ExprOpType::ADD, ExprOpType::SUB }))
+            return false;
+
+        // Deduce net sign of term.
+        const ExpressionTreeNode *parent = node.parent;
+        const ExpressionTreeNode *cur = &node;
+        int polarity = 1;
+
+        while (parent && isOpCode(*parent, { ExprOpType::ADD, ExprOpType::SUB })) {
+            if (parent->op == ExprOpType::SUB && cur == parent->right)
+                polarity = -polarity;
+
+            cur = parent;
+            parent = parent->parent;
+        }
+
+        ++origNumTerms;
+        expr.addTerm(node.valueNum, polarity);
+        index[node.valueNum] = &node;
+        return true;
+    });
+
+    expr.expand(index);
+    bool canonical = expr.canonicalize(index);
+
+    if (expr.numTerms() < origNumTerms || !canonical) {
+        ExpressionTreeNode *seq = expr.emit(tree, index);
+        replaceNode(node, *seq);
+        return true;
+    }
+
+    return false;
+}
+
+bool analyzeMultiplicativeExpression(ExpressionTree &tree, ExpressionTreeNode &node)
+{
+    std::unordered_map<int, const ExpressionTreeNode *> index;
+
+    std::vector<int> origOrder;
+    ExponentMap expr;
+    size_t numDivs = 0;
+
+    node.preorder([&](ExpressionTreeNode &node)
+    {
+        if (node.op == ExprOpType::DIV)
+            ++numDivs;
+
+        if (isOpCode(node, { ExprOpType::MUL, ExprOpType::DIV }))
+            return false;
+
+        // Deduce net sign of term.
+        const ExpressionTreeNode *parent = node.parent;
+        const ExpressionTreeNode *cur = &node;
+        int polarity = 1;
+
+        while (parent &
```

Comment:
Leftover debug code? Or is this some deep magic?

Reply:
Thanks. Fixed.

Reply:
If you think you resolved the issue.

Click on the "Resolve Conversation" button. Neither kblomster, nor anyone else afaik can do so.

If this turns out to be wrong, one can always start a new review. :)